xla/service/gpu/gpu_performance_model.cc

```diff
@@ -299,5 +382,174 @@ void GpuPerformanceModel::RecordEstimatedRunTime(
+             ? sm60_nvlink_bw
+             : sm80_nvlink_bw;
+}
+
+/*static*/ absl::Duration
+GpuPerformanceWithCollectiveModel::ComputeAllreduceTime(
+    const HloInstruction& instr, const GpuHloCostAnalysis* cost_analysis,
+    const GpuDeviceInfo& gpu_device_info) {
+  // We use nccl group call to launch multiple allreduces so launch overhead
+  // only occurs once.
+  absl::Duration total_time = kKernelLaunchOverhead;
+  auto* cc_ptr = std::get_if<stream_executor::CudaComputeCapability>(
+      &gpu_device_info.compute_capability);
+  CHECK(cc_ptr != nullptr);
+
+  stream_executor::CudaComputeCapability compute_cap = {cc_ptr->major,
+                                                        cc_ptr->minor};
+  int64_t size_of_speed_array = intra_node_speeds.size();
+  int64_t size_of_sm90_speed_array = intra_node_speeds_sm90.size();
+
+  int num_speeds = compute_cap.major >= se::CudaComputeCapability::HOPPER &&
+                           compute_cap.minor >= 0
+                       ? size_of_sm90_speed_array
+                       : size_of_speed_array;
+  const double* speeds =
+      compute_cap.major >= se::CudaComputeCapability::HOPPER &&
+              compute_cap.minor >= 0
+          ? intra_node_speeds_sm90.data()
+          : intra_node_speeds.data();
+
+  int speed_index = 0;
+  float max_sys_bw =
+      GetMaxSysBwFromGpu(compute_cap, low_latency_max_bandwidths.data());
+
+  CHECK_GT(max_sys_bw, 0);
+
+  while ((speed_index < num_speeds - 1) && speeds[speed_index] > max_sys_bw) {
+    speed_index++;
+  }
+  float bw_intra_node = speeds[speed_index];
+  int64_t num_devices = cost_analysis->NumOfDevices(instr);
+
+  int64_t min_nchannels =
+      std::max(num_devices, GetMinNumberOfChannels(CollectiveAlgo::RING));
+  int64_t num_channels =
+      std::max(min_nchannels, GetNcclMaxNumChannels(CollectiveAlgo::RING));
+  int default_threads =
+      (bw_intra_node * num_channels <= pci_bw) ? 256 : ll128_nthreads;
+
+  int warp_size = gpu_device_info.threads_per_warp;
+  int num_threads = GetNumThreads(warp_size, ll128_nthreads / 4, ll128_nthreads,
+                                  default_threads);
+
+  // Since channels are pipelined together, compute time will only occur as in a
+  // single channel.
+  absl::Duration compute_time_per_channel =
+      ComputeTime(gpu_device_info,
+                  cost_analysis->flop_count(instr) / num_channels, num_threads);
+  total_time += compute_time_per_channel;
+
+#if GOOGLE_CUDA
+  // We will use nvml library to detect nvlink capability
+  // to see if it supports p2p communication.
+  // We first load libnvidia-ml.so and assign symbols to function pointers
+  // to avoid linking errors.
+  // Then gpu 0 will be used to query for nvlink capability, note that
+  // we only look at link 0 of gpu 0 since all other links are assumed
+  // to have the same capability.
+  void* libhandle = dlopen("libnvidia-ml.so.1", RTLD_NOW);
```

Comment:
This potentially fragile mechanism to load the library probably deserves a test case at least for easier debugging of search and loading.

Reply:
makes sense, I separated all the nvml code into individual functions and i added a test in gpu_performance_model_test to test those functions.